master/internal/resourcemanagers/resource_pool.go

```diff
@@ -344,12 +356,42 @@ func (rp *ResourcePool) Receive(ctx *actor.Context) error {
 func (rp *ResourcePool) receiveAgentMsg(ctx *actor.Context) error {
 	switch msg := ctx.Message().(type) {
 	case sproto.AddAgent:
-		ctx.Log().Infof("adding agent: %s", msg.Agent.Address().Local())
+		agentID := msg.Agent.Address().Local()
+		ctx.Log().Infof("adding agent: %+v", agentID)
 		rp.agents[msg.Agent] = true
+		err := rp.db.AddAgent(&model.AgentStats{
+			ResourcePool: rp.config.PoolName,
+			AgentID:      &agentID,
+			Slots:        msg.Slots,
+			StartTime:    time.Now().UTC().Truncate(time.Millisecond),
+		})
+		if err != nil {
+			ctx.Log().WithError(err)
+		}
 
 	case sproto.RemoveAgent:
-		ctx.Log().Infof("removing agent: %s", msg.Agent.Address().Local())
+		agentID := msg.Agent.Address().Local()
+		ctx.Log().Infof("removing agent: %s", agentID)
 		delete(rp.agents, msg.Agent)
+		endTime := time.Now().UTC().Truncate(time.Millisecond)
+		err := rp.updateAgentEndStats(agentID, &endTime)
+		if err != nil {
+			ctx.Log().WithError(err)
+		}
+		if rp.provisioner != nil {
+			ctx.Tell(rp.provisioner, sproto.EndInstanceStats{})
```

Comment:
can the provisioner stats tracking be done exclusively in the provisioner itself, without resource pool?

Reply:
done